src/context.c

```diff
@@ -290,6 +291,189 @@ ly_ctx_new(const char *search_dir, uint16_t options, struct ly_ctx **new_ctx)
     return rc;
 }
 
+static LY_ERR
+ly_ctx_new_yl_legacy(struct ly_ctx *ctx, struct lyd_node *yltree)
+{
+    struct lyd_node *module, *node;
+    struct ly_set *set;
+    const char **feature_arr = NULL;
+    const char *name = NULL, *revision = NULL;
+    struct ly_set features = {0};
+    ly_bool imported = 0;
+    const struct lys_module *mod;
+    LY_ERR ret = LY_SUCCESS;
+
+    LY_CHECK_RET(ret = lyd_find_xpath(yltree, "/ietf-yang-library:yang-library/modules-state/module", &set));
+
+    /* process the data tree */
+    for (uint32_t i = 0; i < set->count; ++i) {
+        module = set->dnodes[i];
+
+        /* initiate */
+        revision = NULL;
+        name = NULL;
+        imported = 0;
+
+        LY_LIST_FOR(lyd_child(module), node) {
+            if (!strcmp(node->schema->name, "name")) {
+                name = LYD_CANON_VALUE(node);
+            } else if (!strcmp(node->schema->name, "revision")) {
+                revision = LYD_CANON_VALUE(node);
+            } else if (!strcmp(node->schema->name, "feature")) {
+                LY_CHECK_GOTO(ret = ly_set_add(&features, node, 0, NULL), error);
+            } else if (!strcmp(node->schema->name, "conformance-type") &&
+                    !strcmp(LYD_CANON_VALUE(node), "import")) {
+                /* imported module - skip it, it will be loaded as a side effect
+                 * of loading another module */
+                imported = 1;
+                break;
+            }
+        }
+
+        if (imported) {
+            continue;
+        }
+
+        feature_arr = malloc(sizeof *feature_arr * (features.count + 1));
```

Comment:
Just a note, you can keep it this way, but the custom is to first write number of items and then the size of each item, you wrote it the other way around. If you stick to this rule, it will follow the way `calloc` has its arguments and it will avoid possible confusion with `*` since the `sizeof` is always at the end.